go/cloud-query/api/proto/cloudquery.proto

```diff
@@ -0,0 +1,74 @@
+syntax = "proto3";
+
+// Determines the package name for the generated code.
+package cloudquery;
+
+// Configures where the generated code will be placed.
+option go_package = "internal/proto/cloudquery";
+
+message AwsCredentials {
+  string access_key_id = 1;
+  string secret_access_key = 2;
+}
+
+message AzureCredentials {
+  string subscription_id = 1;
+  string tenant_id = 2;
+  string client_id = 3;
+  string client_secret = 4;
+}
+
+message GcpCredentials {
+  string impersonation_token = 1;
+}
+
+message Connection {
+  string provider = 1;
+  oneof credentials {
+    AwsCredentials aws = 2;
+    AzureCredentials azure = 3;
+    GcpCredentials gcp = 4;
+  }
+}
+
+message QueryInput {
+  Connection connection = 1;
+  string query = 2;
+}
+
+message SchemaInput {
+  Connection connection = 1;
+  optional string query = 2;
+}
+
+message ExtractInput {
+  Connection connection = 1;
+}
+
+message QueryOutput {
+  repeated string columns = 1;
+  map<string, string> result = 2;
```

Comment:
is there a way to have the value type be anything outside of string here?  Could be nice if we want to put real logic on these outputs

Reply:
It's not really possible since protobuf does not support any type. It has to be a known type in the end. The closest thing here is string where complex values/objects are JSON encoded strings. That is what we are doing now. Column types can be found using schema api and it would be on the client to properly parse it.